pyvkfft/opencl.py

```diff
@@ -155,20 +155,58 @@ def _make_config(self):
                                          int(self.use_lut), int(self.keepShaderCode),
                                          n_batch, skipx, skipy, skipz)
 
-    def fft(self, src: cla.Array, dest: cla.Array = None):
+    def fft(self, src: cla.Array, dest: cla.Array = None, queue: cl.CommandQueue = None):
         """
         Compute the forward FFT
         :param src: the source pyopencl Array
         :param dest: the destination pyopencl Array. Should be None for an inplace transform
+        :param queue: the pyopencl CommandQueue to use for the transform. If not given,
+            the queue given when constructing the application is used.
         :raises RuntimeError: in case of a GPU kernel launch error
         :return: the transformed array. For a R2C inplace transform, the complex view of the
             array is returned.
         """
+        if not queue:
+            if dest is None:
+                if src.queue is None:
+                    warnings.warn("queue is not given and the source array does not have a queue "
+                        "attached to it. Falling back to the queue given to the constructor "
+                        "of VkFFTApp. This is deprecated and will stop working in the future. "
+                        "Use src_array.with_queue(queue) to attach a queue to the array or "
+                        "pass a queue to this method", DeprecationWarning)
+                    queue = self.queue
+                else:
+                    queue = src.queue
+            elif dest.queue is None and src.queue is None:
+                warnings.warn("queue is not given and the source/dest arrays do not have a queue "
+                        "attached to it. Falling back to the queue given to the constructor "
+                        "of VkFFTApp. This is deprecated and will stop working in the future. "
+                        "Use src_array.with_queue(queue) to attach a queue to the array or "
+                        "pass a queue to this method", DeprecationWarning)
+                queue = self.queue
+            elif dest.queue != src.queue:
+                if dest.queue is None:
+                    queue = src.queue
+                elif src.queue is None:
+                    queue = dest.queue
+                else:
+                    warnings.warn("queue is not given and the source/dest arrays are not equal. "
+                        "Falling back to the queue given to the constructor. This is deprecated "
+                        "and will stop working in the future. "
+                        "Use src_array.with_queue(queue) to attach a queue to the array or "
+                        "pass a queue to this method", DeprecationWarning)
+                    queue = self.queue
+            else:
+                queue = src.queue
+
+        if queue != self.queue:
+            self.queue.finish()
```

Comment:
I don't think it's a good idea to include a finish() here , as it would break asynchronous execution.
It may create issues but anyone juggling multiple queues should manage synchronisation explicitly.